### PR TITLE
WIP: Issue/368

### PR DESCRIPTION
**Pull request by vigneshsarma**
Refer: #368

### CODE REVIEW

File: src/snippet/mod.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    use tokenizer::{SimpleTokenizer, box_tokenizer};
+    use std::iter::Iterator;
+    use std::collections::BTreeMap;
+    use super::{search_fragments, select_best_fragment_combination};
+
+    #[test]
+    fn test_snippet() {
+        let tokenizer = SimpleTokenizer;
+
+        let t = box_tokenizer(tokenizer);
```
Comment (fulmicoton) [inline src/snippet/mod.rs]: let's avoid one letter variables. 
s/t/tokenizer/g

File: src/snippet/mod.rs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+impl Snippet {
+
+    /// Returns a hignlightned html from the `Snippet`.
+    pub fn to_html(&self) -> String {
+        let mut html = String::new();
+        let mut start_from: usize = 0;
+
+        for item in self.highlighted.iter() {
+            html.push_str(&self.fragments[start_from..item.start]);
```
Comment (fulmicoton) [inline src/snippet/mod.rs]: Please encode the html entities in the original text. You can add [htmlescape](https://crates.io/crates/htmlescape) as a dependency.

File: src/snippet/mod.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+                          start_offset: start_offset,
+                          stop_offset: end_offset,
+                          num_chars: 0,
+                          highlighted: vec![]}
+    }
+
+    /// Updates `score` and `highlighted` fields of the objects.
+    ///
+    ///
+    fn calculate_score(&mut self, token: &Token, terms: &BTreeMap<String, f32>) {
```
Comment (fulmicoton) [inline src/snippet/mod.rs]: this should be called add_token.

Comment (fulmicoton) [inline src/snippet/mod.rs]: This should be called try_add_token`.

The comment above should explain that if the token is one of the keyword, then the fragment length will be extended.

File: src/snippet/mod.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    let mut fragments:Vec<FragmentCandidate> = vec![];
+
+    loop {
+        if let Some(next) = token_stream.next() {
+            if (next.offset_to - fragment.start_offset) > max_num_chars {
+                let txt = &text[fragment.start_offset..fragment.stop_offset];
+                if fragment.score > 0.0 {
+                    fragments.push(fragment)
+                };
+                fragment = FragmentCandidate::new(next.offset_from, next.offset_to);
```
Comment (fulmicoton) [inline src/snippet/mod.rs]: I think there is a bug if `next` is a keyword. Then it never contributes to the score does it.

File: src/tokenizer/tokenizer.rs [hunk 5]
```diff
@@ -130,7 +130,7 @@ where
     }
 }
 
-pub(crate) fn box_tokenizer<A>(a: A) -> Box<BoxedTokenizer>
+pub fn box_tokenizer<A>(a: A) -> Box<BoxedTokenizer>
```
Comment (fulmicoton) [inline src/tokenizer/tokenizer.rs:133]: why remove `(crate)`?

File: src/snippet/mod.rs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    use std::collections::BTreeMap;
+    use super::{search_fragments, select_best_fragment_combination};
+
+    #[test]
+    fn test_snippet() {
+        let tokenizer = SimpleTokenizer;
+
+        let t = box_tokenizer(tokenizer);
+
+        let text = "Rust is a systems programming language sponsored by Mozilla which describes it as a \"safe, concurrent, practical language\", supporting functional and imperative-procedural paradigms. Rust is syntactically similar to C++[according to whom?], but its designers intend it to provide better memory safety while still maintaining performance.
```
Comment (fulmicoton) [inline src/snippet/mod.rs]: That's a good unit test. Keep it. Can you add more simpler unit test (the words can be one letter words) that explore the specification.

For instance :
- What happens if the first token is a keyword?
- if the text is empty?
- if the text contains no keyword?

(It's ok if the current specification is not great. We'll fix the unit tests as we improve the code.)

**Comment by fulmicoton** [comment]
Thanks. I added some commetns to the CR. I understand it might feel a bit confusing without being able to run it with a query. I'll try to fill the gap later this week so we can see the result of your work on a actual queries.

File: src/tokenizer/tokenizer.rs (hunk 5 above)
Comment (vigneshsarma) [reply src/tokenizer/tokenizer.rs:133]: With this, I get the following error when I compile, 

```rust
error[E0364]: `box_tokenizer` is private, and cannot be re-exported
   --> src/tokenizer/mod.rs:156:43
    |
156 | pub use self::tokenizer::{BoxedTokenizer, box_tokenizer};
```

**Comment by vigneshsarma** [comment]
@fulmicoton What do you think of the current PR?
To fix 
> if the term is repeated twice, it will add the score twice.

I think we will have to introduce a separate struct, which maintains the tokens that have been counted already. 
Also shouldn't we use a better comparison than a simple `terms.get()`?

> not calculating num_chars.

I am not completely sure how to fix this.

**Comment by fulmicoton** [comment]
@vigneshsarma I need a little bit more time to review your code, and finishing the "connecting-the-dot" code to make your highlighter usable by users.

Please give me some more time... Overall we should be able to merge somewhere next week.

**Pull request merged**